testgres/pubsub.py

```diff
@@ -0,0 +1,202 @@
+"""
+Unlike physical replication the logical replication allows users replicate only
+specified databases and tables. It uses publish-subscribe model with possibly
+multiple publishers and multiple subscribers. When initializing publisher's
+node ``allow_logical=True`` should be passed to the :meth:`.PostgresNode.init()`
+method to enable PostgreSQL to write extra information to the WAL needed by
+logical replication.
+
+To replicate table ``X`` from node A to node B the same table structure should
+be defined on the subscriber's node as logical replication don't replicate DDL.
+After that :meth:`~.PostgresNode.publish()` and :meth:`~.PostgresNode.subscribe()`
+methods may be used to setup replication. Example:
+
+>>> from testgres import get_new_node
+>>> with get_new_node() as nodeA, get_new_node() as nodeB:
+...     nodeA.init(allow_logical=True).start()
+...     nodeB.init().start()
+...
+...     # create same table both on publisher and subscriber
+...     create_table = 'create table test (a int, b int)'
+...     nodeA.safe_psql(create_table)
+...     nodeB.safe_psql(create_table)
+...
+...     # create publication
+...     pub = nodeA.publish('mypub')
+...     # create subscription
+...     sub = nodeB.subscribe(pub, 'mysub')
+...
+...     # insert some data to the publisher's node
+...     nodeA.execute('insert into test values (1, 1), (2, 2)')
+...
+...     # wait until changes apply on subscriber and check them
+...     sub.catchup()
+...
+...     # read the data from subscriber's node
+...     nodeB.execute('select * from test')
+PostgresNode(name='...', port=..., base_dir='...')
+PostgresNode(name='...', port=..., base_dir='...')
+''
+''
+[(1, 1), (2, 2)]
+"""
+
+from six import raise_from
+
+from .consts import LOGICAL_REPL_MAX_CATCHUP_ATTEMPTS
+from .defaults import default_dbname, default_username
+from .exceptions import CatchUpException
+from .utils import options_string
+
+
+class Publication(object):
+    def __init__(self, name, node, tables=None, dbname=None, username=None):
+        """
+        Constructor. Use :meth:`.PostgresNode.publish()` instead of direct
+        constructing publication objects.
+
+        Args:
+            name: publication name.
+            node: publisher's node.
+            tables: tables list or None for all tables.
+            dbname: database name used to connect and perform subscription.
+            username: username used to connect to the database.
+        """
+        self.name = name
+        self.node = node
+        self.dbname = dbname or default_dbname()
+        self.username = username or default_username()
+
+        # create publication in database
+        t = "table " + ", ".join(tables) if tables else "all tables"
+        query = "create publication {} for {}"
+        node.execute(query.format(name, t), dbname=dbname, username=username)
+
+    def drop(self, dbname=None, username=None):
+        """
+        Drop publication
+        """
+        self.node.safe_psql(
```

Comment:
Could we replace `safe_psql` with `execute`?

Reply:
Yep, I overlooked it